File: set1/detect_single_character_xor/parallel_detect_single_character_xor.py
#!/usr/bin/env python3

################################################################################
# The matasano crypto challenges
# http://cryptopals.com/sets/1/challenges/4/
# Set 1   Challenge 4
# Detect single-character XOR
################################################################################
# One of the 60-character strings in the input file has been encrypted 
# by single-character XOR. Find it. 
# Key:      int=53, char='5'
# Message:  Now that the party is jumping
#
# NOTE: This implementation is parallel
################################################################################

import os
import sys
import string
import logging
import argparse
import multiprocessing
from contextlib import ContextDecorator


################################################################################
logger = logging.getLogger('parallel_detect_single_character_xor')
logger.setLevel(logging.DEBUG)
#logger.setLevel(logging.INFO)
ch = logging.StreamHandler()
ch.setLevel(logging.DEBUG)
formatter = logging.Formatter('%(levelname)s: %(message)s')
ch.setFormatter(formatter)
logger.addHandler(ch)
################################################################################


class FileSplitter(ContextDecorator):
    """
    Splits input file into multiple smaller temporary files (one file per worker)
    The size of a smaller file is determined like this:
        size_of_input_file / number_of_workers

    After the worker is done decoding its file, 
    FileSplitter removes this temporary small file
    """

    def __init__(self, input_file, number_of_workers):
        self._input_file = input_file
        self._number_of_workers = number_of_workers
        self._split

    def __enter__(self):
        self._split_files = self._split()
        for f in self._split_files:
            logger.debug('Tmp file: {0}'.format(f.name))
        return self

    def __exit__(self, *exc):
        for f in self._split_files:
            logger.debug('Cleaning up temporary file: {0}'.format(f.name))
            os.remove(f.name)

        return False

    def __next__(self):
        for f in self._split_files:
            yield f

    def _split(self):
        logger.debug('Splitting file {0}'.format(self._input_file))
        split_files = []
        file_size, split_file_size = self._do_prelim_calculations()
        with open(self._input_file, 'r') as f:
            for worker in range(self._number_of_workers):
                split_files.append(self._read_and_split(f, worker, split_file_size))
        return split_files

    def _read_and_split(self, file_, worker, split_file_size):
        import tempfile
        with tempfile.NamedTemporaryFile(prefix=os.path.basename(self._input_file) + '.' + str(worker) + '.', delete=False) as tmp:
            logger.debug('Temporary file {0} created'.format(tmp.name))
            bytes_read = 0
            for line in file_:
                logger.debug('Writing line [{0}]'.format(line[:len(line) - 1] if line[-1] == '\n' else line))
                tmp.write(bytes(line, 'UTF-8'))
                bytes_read += len(line)
                if bytes_read >= split_file_size:
                    return tmp

            return tmp

    def _do_prelim_calculations(self):
        import math
        # Get file size in bytes
        file_size = os.path.getsize(self._input_file)

        # Calculate the size of each file (in bytes) to be split
        split_file_size = file_size // self._number_of_workers
        logger.debug('File size: {0}; split file size: {1}'.format(file_size, split_file_size))
        return file_size, split_file_size


def find_key(key, tuple_):
    return chr(int(tuple_[0] + tuple_[1], base=16) ^ key)

def decode_with_key(key, s):
    decoded_msg = ''
    for t in zip(s[0::2], s[1::2]):
        decoded_msg += find_key(key, t)

    if len([c for c in decoded_msg if c in string.ascii_letters + ' \n']) == len(decoded_msg):
        logger.info('[*] Trying the key: int: {0}, char: {1}'.format(key, chr(key)))
        logger.info('Decoded message: {0}'.format(decoded_msg))
        return decoded_msg

    return None
    
def remove_eol(s):
    """Removes trailing '\n' if there is one"""
    return s[0:len(s) - 1] if s[len(s) - 1] == '\n' else s

def decode(queue_, file_to_decode):
    results = []
    with open(file_to_decode, 'r') as f:
        for encoded_line in f:
            line = remove_eol(encoded_line)
            #logger.debug('File: {0}\n\tDecoding [{1}]'.format(f.name, line))
            for key in range(0, 256):
                decoded_msg = decode_with_key(key, line)
                if decoded_msg is not None:
                    results.append((f.name, key, decoded_msg))

    queue_.put(results)

def main():
    parser = argparse.ArgumentParser(description='Parallel detect single-character XOR')
    parser.add_argument('-w', '--workers', dest='workers', default=1, type=int,
            help='Number of workers to split the decoding work amongst')
    parser.add_argument('file', type=str, help='File to decode')
    args = parser.parse_args()

    with FileSplitter(args.file, args.workers) as fs:
        proc = []
        q = multiprocessing.Queue()
        for f in next(fs):
            #logger.info('Decoding file {0}'.format(f.name))
            proc.append(multiprocessing.Process(target=decode, args=(q, f.name,)))
            proc[-1].start()

        for p in proc:
            res = q.get()
            if res:
                logger.info('Results: {0}'.format(res))
            p.join()

if __name__ == '__main__':
    main()

File: set1/detect_single_character_xor/parallel_detect_single_character_xor.ml
(*******************************************************************************
* The matasano crypto challenges
* http://cryptopals.com/sets/1/challenges/4/
* Set 1   Challenge 4
* Detect single-character XOR
* Compile it like this:
* $ ocamlfind ocamlc -package batteries,core,forkwork -thread -linkpkg 
*   -o parallel_detect_single_character_xor parallel_detect_single_character_xor.ml
********************************************************************************
* One of the 60-character strings in the input file has been encrypted 
* by single-character XOR. Find it. 
* Key:      int=53, char='5'
* Message:  Now that the party is jumping
*
* NOTE: This implementation is parallel (requires ForkWork package)
*******************************************************************************)

open Core.Std
open Batteries

let tmp_files = RefList.empty ()
let _workers = ref 1


(*******************************************************************************
 * FILE SPLITTING LOGIC - BEGIN 
 ******************************************************************************)
let prelim_calculations number_of_workers filename =
    (* Calculate total file size and split file size
     * (total_file_size / number_of_workers) in bytes
     *)
    let total_file_size = File.size_of filename in
    total_file_size / number_of_workers

let line_stream_of_channel channel =
    Stream.from
      (fun _ ->
         try Some (input_line channel) with End_of_file -> None)

let rec write_line out stream bytes_read_so_far split_file_size =
    if bytes_read_so_far < split_file_size then
        let line = (Stream.next stream) ^ "\n" in
        String.print out line;
        write_line out stream (bytes_read_so_far + (String.length line)) split_file_size

let make_tmp stream split_file_size =
    File.with_temporary_out ~suffix:".tmp"
        (fun out name -> RefList.add tmp_files name; write_line out stream 0 split_file_size)

let rec process_stream stream workers split_file_size =
    if workers > 0 then begin
        make_tmp stream split_file_size;
        ignore (process_stream stream (workers - 1) split_file_size)
    end

let read_file workers filename () =
    let in_chan = open_in filename in
    let split_file_size = prelim_calculations workers filename in
    let stream = line_stream_of_channel in_chan in
    try
        _workers := workers;
        process_stream stream !_workers split_file_size;
        close_in in_chan
    with e ->
        (*RefList.iter (fun f -> print_endline(f)) tmp_files;*)
        close_in in_chan

(*******************************************************************************
 * FILE SPLITTING LOGIC - END 
 ******************************************************************************)


(*******************************************************************************
 * DECODING LOGIC - BEGIN
 ******************************************************************************)

let normalize s =
    let make_hex_str i (*index*) =
        let prev = String.get s (i - 1)
        and cur  = String.get s i in
        ("0x" ^ String.make 1 prev) ^ (String.make 1 cur)
    in
    let pair_elem i ch =
        if i mod 2 <> 0 && i <> 0 then
            let hex = int_of_string (make_hex_str i) in char_of_int hex
        else
            Char.chr 0
    in
    BatString.filter ((<>) (Char.chr 0)) (BatString.mapi pair_elem s)

let int_from_hex_char ch =
    BatInt.of_string ("0x" ^ (BatString.make 1 ch))

let xor_chars a b =
    Char.chr ((Char.code a) lxor b)

let is_readable msg = 
    (* Determines if msg is readable 
     * (i.e., consists of printable ascii characters)
     *)
    let r1 = BatChar.range ' '  ~until:'~'
    and r2 = BatChar.range '\n' ~until:'\n' in
    let merged = List.of_enum (BatEnum.merge (fun _ _ -> true) r1 r2) in
    BatEnum.for_all (fun c -> List.mem c merged) (BatString.enum msg)

let rec decode key s results filename =
    if key < 256 then decode_with_key key s results filename
and decode_with_key key s results filename =
    let decoded_msg = BatString.map (fun ch -> xor_chars ch key) s in
    let readable = is_readable decoded_msg in
        if readable then 
            begin
                (*Printf.printf "[*] Trying the key: %d, char: %C\nDecoded message: %s\n" key (Char.chr key) decoded_msg;*)
                RefList.add results (decoded_msg, key, filename);
                decode (key + 1) s results filename
            end
        else
            decode (key + 1) s results filename

let decode_file filename =
    let filelines = File.lines_of filename in
    let results = RefList.empty () in
    Enum.iter (fun line -> (*Printf.printf "File: %s; Decoding [%s]\n" filename line;*) decode 0 (normalize line) results filename) filelines;
    RefList.to_list results

(*******************************************************************************
 * DECODING LOGIC - END
 ******************************************************************************)


let command =
    Command.basic
        ~summary:"Parallel detect single-character XOR - OCaml implementation"
        ~readme:(fun () -> "Matasano crypto challenge: set 1, challange 4")
        Command.Spec.(
            empty 
            +> flag "-w" (optional_with_default 1 int) ~doc:"Number of workers"
            +> anon ("filename" %: string))
        read_file

let print_results results = 
    List.iter (fun (m, k, f) -> Printf.printf "Possible decoded message (key=%C, file=%s):\n%s\n" (Char.chr k) f m) results

let split_work workers =
    Printf.printf "Number of workers: %d\n" workers;
    let list_of_files = RefList.to_list tmp_files in
    let results = ForkWork.map_list decode_file list_of_files in
    print_results (List.flatten results)


let () = 
    (* Step #1: Split input file into multiple smaller files *)
    Command.run ~version:"1.0" ~build_info:"RWO" command;
    (* Step #2: Spawn one worker process per file and collect the results *)
    split_work !_workers 
